Charts

dcbr opened the pull request:
### Overview

Hi there pyvista devs,

I have been messing around with the vtk Charts API lately and thought it would make a great addition to pyvista.
This is most definitely a WIP, as I have not added any examples/tests/documentation and many plot & chart features of vtk can still be added. However I wanted to show you this prototype already to see whether you would like to see this added to pyvista or not. If so, I can further work on this in my free time. Of course some help/advice from your side would also be very much appreciated! :)


### Details

- Expose VTK Charts API using a pythonic, matplotlib-like interface
- Plotting module allows to easily combine 2D charts (as overlays) and 3D meshes/actors in the same renderer (viewport). Interaction with the charts or 3D scene can be toggled through a right click.
- Support to integrate matplotlib figures into a pyvista plotting window

### Prototype

Two chart overlays (left: matplotlib figure, right: vtkChartXY) on the same renderer --> see 'chart_overlays.py' example.

https://user-images.githubusercontent.com/15089458/123146422-1fc5c080-d45e-11eb-99a3-b73dc178766f.mp4


### Task list:

- [x] Chart2D: Implementation of scatter, line, area, bar & stacked plots
- [x] ChartPie & ChartBox: Implementation of pie & box plots
- [x] Chart3D: Implementation of Line, Scatter & Volume plots
- [x] ChartMPL: Matplotlib support
- [x] Examples
- [x] Tests
- [x] Documentation
- [x] Fix remaining TODOs

### Known issues:

- ~~No support for vtk>=9.0 until [issue 18122](https://gitlab.kitware.com/vtk/vtk/-/issues/18122) is fixed (and included in pypi package)~~ OK for latest pypi wheels
- ~~Misplaced background color for vtk charts, see [issue 18238](https://gitlab.kitware.com/vtk/vtk/-/issues/18238)~~ Workaround available for vtk>=9.0
- ChartPie legend not fully visible when changing the plot's label or drawn behind the chart, requires VTK fix (will be available in VTK 9.2).
- ChartBox plots and log scale axis not correctly drawn in docs (offscreen rendering), requires VTK fix (will be available in VTK 9.2).

Note: Only tested on my local pc (Windows 10, python 3.7, vtk 9.0.3)

akaszynski commented:
This is really cool!

Thanks for working on this, I look forward to reviewing and having it out in our next release.

tkoyama010 commented:
Awesome! I was impressed in your job.

akaszynski commented on examples/02-plot/chart_basics.py:
```diff
@@ -0,0 +1,129 @@
+
+import pyvista as pv
+import numpy as np
+RNG = np.random.default_rng(1)
```
Let's avoid having all cap variables.  These are generally used for [constants](https://www.python.org/dev/peps/pep-0008/#constants)

adeak replied:
It's also numpy custom to just call this `rng`. I support use of the new random API!

dcbr replied:
ok, I will change that. Seeding was mainly to get some consistent plots while trying out things, but can definitely be removed in the final version.

akaszynski replied:
Actually, I think seeding it is fine.  In `numpydoc` with their plot directive, they always seed their random examples to generate consistent plots.  I think we should do the same.

adeak replied:
There's a trade-off between reproducibility and best practices. I tend to prefer setting a good example unless there's a very good reason not to. Not teaching users that seeding generators is normal is important. Getting the exact same plot is not that important to me as long as the overall look is the same and there are no edge cases that can break the example.

This is just to explain my stance. We can go ahead and seed regardless.

akaszynski replied:
I'm either way for seeding.  @dcbr, I'll leave it up to you.

dcbr replied:
Ok, I can understand both arguments pro and con. My personal view is slightly more in favour of seeding though. In these examples, I see the rng calls more as a shorthand notation to quickly get some interesting data instead of having to hardcode some arbitrary datasets yourself. The actual randomness/uniqueness of this generated data is then less important than the consistency imo. I could add a short comment to the example to inform the user why seeding is used:
`rng = np.random.default_rng(1)  # Seeded random number generator for consistent data generation`

akaszynski commented on pyvista/plotting/charts.py:
```diff
@@ -150,6 +201,20 @@ def width(self):
 
+    @property
+    def pen(self):
+        """
+        :return: Retrieve vtkPen object controlling how lines in this plot are drawn.
+        """
```
We're using the numpydoc documentation style

```suggestion
        """vtkPen object controlling how lines in this plot are drawn.

        Returns
        --------
        vtk.vtkPen
            `vtkPen` object controlling how lines in this plot are drawn.

        Examples
        ---------
        >>> import pyvista
        """
```
All public methods should have documented methods, and if possible, include plots.  These are now automatically rendered (you may have to merge with main).  These are great for demonstrating functionally.  For example:
https://dev.pyvista.org/api/core/_autosummary/_autosummary/pyvista.PolyDataFilters.boolean_difference.html

dcbr replied:
Ok, that's really cool, will do that! I will focus more on the documentation in future commits, once the main implementation is done and the interface is +- fixed

akaszynski commented on examples/02-plot/chart_overlays.py:
```diff
@@ -0,0 +1,56 @@
+"""
```
Use sphinx references to allow for the automatic generation of links between the rendered documentation.

```suggestion
"""
.. _chart_overlays_example:
```

akaszynski commented on examples/02-plot/chart_overlays.py:
```diff
@@ -0,0 +1,56 @@
+
+
+This example shows how you can combine multiple charts as overlays in the same renderer. For an overview of the
+different chart types you can use, please refer to `chart_basics`.
```
```suggestion
different chart types you can use, please refer to :ref:`chart_basics_example`.
```

akaszynski commented on examples/02-plot/chart_basics.py:
```diff
@@ -0,0 +1,129 @@
+"""
```
```suggestion
"""
.. _chart_basics_example:
```

akaszynski commented on examples/02-plot/chart_basics.py:
```diff
@@ -0,0 +1,129 @@
+
+
+This example shows how different types of charts can be added to the scene. A more complex example, showing how to
+combine multiple charts as overlays in the same renderer, is given in `chart_overlays`.
```
```suggestion
combine multiple charts as overlays in the same renderer, is given in :ref:`chart_overlays_example`.
```

akaszynski commented on pyvista/core/objects.py:
```diff
@@ -47,10 +47,18 @@ def __init__(self, *args, **kwargs):
+        np_table = self._prepare_arrays(arrays)
         for i, array in enumerate(np_table):
             self.row_arrays[f'Array {i}'] = array
         return
```
Unnecessary return here and in other spots.  Just noticed it and it's not part of this PR, just figured to clean it up.
```suggestion
```

dcbr commented on pyvista/plotting/charts.py:
```diff
@@ -0,0 +1,827 @@
+        return obj
+
+
+class _vtkWrapper(object, metaclass=_vtkWrapperMeta):
```
Thanks for your comments, I will try to take these things into account!
I had another question regarding the wrapping of vtk objects. From what I have seen in the pyvista codebase, the vtk type is typically subclassed and the wrapped instance is passed around to the vtk methods. This works fine and I tried to follow this approach throughout this PR. But what should I do when I have an existing vtk object (created internally by the vtk library within a function call) and want to wrap it to add certain functionality? Is this already done somewhere within the pyvista codebase?
I noticed you can create a wrapped version of native vtk objects in the following way:
```py
obj = _vtk.vtkPlotPie()
wrapped = _vtk.vtkPlotPie(obj.__this__)
assert obj.__this__ == wrapped.__this__
```
Unfortunately, this doesn't seem to work for subclassed types. E.g. the following does not properly wrap the original object
```py
class PiePlot(_vtk.vtkPlotPie):
   pass
obj = PiePlot()
wrapped = PiePlot(obj.__this__)
assert obj.__this__ == wrapped.__this__  # Not correctly wrapping obj
```
For now I came up with this wrapping code using some metaclass and getattribute magic, but not sure whether this is the best approach...

adeak replied:
I had the same problem with the [`Light`](https://github.com/pyvista/pyvista/blob/main/pyvista/plotting/lights.py#L30) class. What I ended up doing was to [define a `classmethod`](https://github.com/pyvista/pyvista/blob/main/pyvista/plotting/lights.py#L1088-L1119) that makes a copy of the input vtk object and throws it away. As far as I can tell this is the price we have to pay for not using composition.

Related PR was https://github.com/pyvista/pyvista/pull/1040 though I can't see too much discussion of this detail there.

dcbr replied:
Ok, I see. But I'm afraid a copy won't be working in this case, as the containing chart (vtkCharPie instance) still has a reference to the original vtkPlotPie object and uses this original object to draw the plot to the screen. So any changes a user makes on that copied instance, won't be reflected on the screen.
So for now, I'll keep the wrapping code. I will create a PR for the vtk codebase to allow us to (re)set the default generated vtkPlotPie object, such that we don't have to do this wrapping (and can pass a new wrapped PiePlot instance to the containing chart). It's weird that they didn't implement this already, as the exact same situation arises for boxplots but there they added a public method to replace the contained plot object...

akaszynski replied:
Another approach is what we used for the [RenderWindowInteractor](https://github.com/pyvista/pyvista/blob/abe056bbc97667c722eb50329cae90ac76fc253c/pyvista/plotting/render_window_interactor.py) where we assign the vtk object as an internal attribute and then wrap that.  It's not perfect, but it works, doesn't require a copy, etc.

adeak replied:
That would be the composition alternative (as opposed to inheritance). I usually prefer that, and I think that's what I wanted to do for `Light` originally ;)

dcbr commented on pyvista/plotting/plotting.py:
```diff
@@ -3770,7 +3775,7 @@ def __init__(self, off_screen=None, notebook=None, shape=(1, 1),
         def on_timer(iren, event_id):
             """Exit application if interactive renderer stops."""
-            if event_id == 'TimerEvent':
+            if event_id == 'TimerEvent' and self.iren._style != "Context":
```
I had to add this when the vtkContextStyleInteractor is active, as otherwise the plotting window was closing immediately. Does this make sense?

akaszynski replied:
That's fine with me.  This might lead to issues with animation on Windows as it's quite sensitive to `on_timer` and `update`.  I'll test this on my windows box before merging to see if animating with a chart works on Windows.  Would be a good test to have.

dcbr replied:
Ok perfect, thanks a lot for your comments and fixes!
By animation, do you mean something like the chart_overlays example or something else? Because that example works well on my machine (Windows 10).

akaszynski replied:
Animation as in your example you posted in the PR overview.  If that works on windows, it will work on Linux.

Really, this is a great PR and will provide a great way to visualize data from within a vtk plot, especially from a data science perspective.  Appreciate your work!

dcbr replied:
Yes the chart_overlays example is based on that example in the overview, so that should be fine.
Glad that you like it! :D

adeak replied:
Yeah, let me echo what Alex said. I haven't had the time to review PRs for a while (or contribute otherwise, really), but the additions in this PR look awesome! And being able to embed a matplotlib axes is a very very pleasant surprise to me.

dcbr reviewed (commented):
So, this took a bit longer than expected, as it was quite tricky to make these charts interactive without causing crashes. A couple of questions/remarks regarding this commit:

- Interaction with a chart is now enabled when the user performs a right click on top of one. Is this ok or should I use another trigger (e.g. double click or shift/ctrl + click)?
- I created a workaround for the background color issues of charts, such that we don't have to wait on official vtk support for that (for which I will submit a PR later). This however requires a vtk version of at least 9.0. Not sure what version you would still like to support? We could maybe fallback to the erroneous behaviour for earlier vtk versions.
- Some 3D charts are now also supported. However I'm really not sure if they add any value. In my opinion they lack a lot of features and are not as customizable as the 2D charts. They also don't really respect the chart's dimensions once you start interacting with them, making it hard to properly integrate in a layout I think. Let me know what your thoughts are regarding this, otherwise I can remove support for them in the next commits.

This more or less finalizes the main implementation, so future commits will focus more on documentation and tests.

dcbr commented on pyvista/plotting/tools.py:
```diff
@@ -364,6 +364,7 @@ def parse_color(color, opacity=None, default_color=None):
     elif len(color) == 3:
         pass
     elif len(color) == 4:
+        opacity = color[3]  # TODO: verify/ask whether this addition doesn't break anything
```
It seems there are also 3 tests failing because of this change (as they expect a list of length 3 as the return value instead of a list of length 4, including the opacity).
In my opinion, it makes sense to also return back the opacity if it was already present in the input argument. This way you don't have to deal with the opacity separately after this call. E.g. you can pass a default opacity value to use if it is not present in the input and use the input's opacity otherwise.
`color = parse_color(color, opacity=1)  # Will always return a list of length 4 with opacity included`
What is your opinion about this?

akaszynski replied:
> In my opinion, it makes sense to also return back the opacity if it was already present in the input argument.

Agreed.  Also, this method lacks docs.  Adding...

dcbr commented:
Ok, I was able to fix the other remaining failing tests caused by my changes (but will wait for your opinion regarding the parse_color method before I commit).
There's however 3 more tests that gave an error on teardown, which I think are not related to the Charts API. Could you have a look at them? I think it's related to renderer attributes that are not cleaned up in the deep_clean() method.

- test_make_movie_teardown failure: caused because renderer's marker_actor attribute (set in add_axes_at_origin method) is not properly cleaned up
- test_multi_renderers_teardown and test_add_remove_floor_teardown: caused because renderer's _bounding_box attribute (set in add_bounding_box method) is not properly cleaned up

akaszynski commented:
> Ok, I was able to fix the other remaining failing tests caused by my changes (but will wait for your opinion regarding the parse_color method before I commit).

Fixed this and added badly needed docs.

> There's however 3 more tests that gave an error on teardown

Cleaned these up.  I'm surprised our unit testing didn't pick up on two of these.

akaszynski commented:
> Interaction with a chart is now enabled when the user performs a right click on top of one. Is this ok or should I use another trigger (e.g. double click or shift/ctrl + click)?

Since left click is the default interaction method for the main renderer, I think thi works.


> I created a workaround for the background color issues of charts, such that we don't have to wait on official vtk support for that (for which I will submit a PR later). This however requires a vtk version of at least 9.0. Not sure what version you would still like to support? We could maybe fallback to the erroneous behaviour for earlier vtk versions.

We have a variety of methods that support only VTK 9.  It's been out for over a year; you can either raise an exception or warning when attempting to use charts on VTK < 9.  I've changed the behavior so charts are only created when needed, so if users don't use it the error/exception won't be raised.

> Some 3D charts are now also supported. However I'm really not sure if they add any value. In my opinion they lack a lot of features and are not as customizable as the 2D charts. They also don't really respect the chart's dimensions once you start interacting with them, making it hard to properly integrate in a layout I think. Let me know what your thoughts are regarding this, otherwise I can remove support for them in the next commits.

This is already a huge PR.  I think we can add that in a followup PR if we think it's needed.

dcbr commented:
Could you have a look at the tests I added for Pen and Brush objects? Is that more or less what you would like to see or should I create comparisons of actual plots to see whether things are actually correctly drawn on screen (and if so, how should I best implement that?)
I added the tests to 'test_plotting', is that ok or should I create a new 'test_charts' file?

akaszynski commented:
> I added the tests to 'test_plotting', is that ok or should I create a new 'test_charts' file?

I'd create a new `test_charts.py` file directly in the `tests` directory that tests the charting functionality without actually plotting it.  Then, I'd create a handful of tests that verify that the plots are generating correctly.

I'll implement this to give you an idea of what it looks like.

dcbr commented:
Ok thanks, then I can continue with the other classes.
To fix issues with the StaticCodeCheck, I removed the type hints for now. I never really worked with type hints before, but thought it would be helpful for users of the module. However it seems trivial stuff like [this](https://github.com/python/mypy/issues/3004) (which I used quite a lot throughout this PR, e.g. for colors) is not properly supported and it doesn't really look like it will be fixed anytime soon. Is it ok to just specify types/descriptions in the docs?

akaszynski commented:
> To fix issues with the StaticCodeCheck, I removed the type hints for now

Agree with that.  It's a bit of a pain to get right, and it's best done incrementally.

> Is it ok to just specify types/descriptions in the docs?

Yes, in fact we do that for a majority of our methods/classes.

akaszynski commented:
Added a variety of documentation in https://github.com/pyvista/pyvista/pull/1432/commits/af2345f3b6d6e2b0f61ff5465630c499a999d5df and https://github.com/pyvista/pyvista/pull/1432/commits/7b60ad830921f0966de365c3cc579072a3fadcce, as well as adding a `show` method for the `Chart2D` class.

Documentation and examples still need quite a bit of work, but we should have all the classes and methods documented with examples:
![tmp](https://user-images.githubusercontent.com/11981631/127787001-e5f8fc22-53e6-4083-b3d7-3e4f96410222.png)

We're using a modified sphinx template to plot the auto-summaries for the classes, and we also have a plot directive that generates figures from any plots we generate.

dcbr commented:
Just FYI I'll be on holiday from the end of next week on, so I won't be adding anything here until the start of September. I'll try to complete all documentation before I leave, but there will probably still be some tests left to implement (which I will of course continue working on afterwards).

akaszynski commented:
> Just FYI I'll be on holiday from the end of next week on, so I won't be adding anything here until the start of September. I'll try to complete all documentation before I leave, but there will probably still be some tests left to implement.

That's no problem!  Great work with all the examples!  There are some tweaks to be made regarding doc style, but overall the examples look spot-on.

akaszynski commented:
Updated the docstrings to remove the `@setter` docstring and moved example to the property.  Sphinx renders only the property method.  Also, you have to have a space between any text you wish to be "plain text" and the actual code to get it to render correctly.

Additionally, I've moved the `matplotlib` import to within the `ChartMPL` class as there's no reason to import it until using the class.  This speeds up the import of pyvista.

dcbr commented:
Sorry for the delay, things got a little busy and there were some bugs I encountered while writing the examples that were quite tricky to handle. I will gradually be adding the remaining docs and tests in the coming weeks.

akaszynski commented:
> Sorry for the delay, things got a little busy and there were some bugs I encountered while writing the examples that were quite tricky to handle. I will gradually be adding the remaining docs and tests in the coming weeks.

No problem!  We got serious with the docs for last release and the CI will be sure to complain at you until it's corrected.  Ping me if you'd like me to take a look at it or help out.

dcbr commented:
@akaszynski Could you make some suggestions on how to fix the remaining (3) stylechecks for the docs, I'm not sure I fully understand why they are failing.

adeak commented on pyvista/plotting/charts.py:
```diff
@@ -0,0 +1,3375 @@
+        return self._scene.GetRenderer()
+
+    def _render_event(self, *args, **kwargs):
+        """This method will be called before every render."""
```
The problem here is that docstring excerpt lines should be phrased as "Do the thing". Instead you have something very different. You should say someting about the functionality of the function on this line, and then mention it being called each time on a separate row.

adeak commented on pyvista/plotting/charts.py:
```diff
@@ -0,0 +1,3375 @@
+        self._renderer = renderer
+
+    def _setup_scene(self):
+        """Setup a new context scene and actor for these charts."""
```
I think the style error here might ba false positive. See if changing to "Set up..." helps:
```suggestion
        """Set up a new context scene and actor for these charts."""
```
I suspect the linter thinks "Setup" is a noun here.

adeak commented on pyvista/plotting/charts.py:
```diff
@@ -0,0 +1,3375 @@
+        return self._add_plot("stack", x, ys, colors=colors, labels=labels)
+
+    def plots(self, plot_type=None):
+        """Return all plots (of the specified type) in this chart.
```
I assume the linter complains about "Return...", thinking that you're specifying inputs and outputs on this line. Arguably a false positive. Perhaps you could use "Retrieve" here as well to make it happy.

But `plot_type` is undocumented, consider adding that. Although I can see that all the others are missing too, from a glance, so that might just be a WIP.

dcbr replied:
All right thanks for the feedback! I'm going to focus on the tests and remaining todos now, then I'll do a final pass through all the changes to fix the last missing parts :)

dcbr replied:
Hmmm, "Retrieve" didn't do the trick, same error message :/

dcbr replied:
Ok, so apparently it's the parenthesis that are causing this. Although I quite liked them because the parameter is optional.

adeak replied:
Ehh, that's dumb of the linter then. There's probably a way to make an exception for that docstring, but I guess the line makes sense even without parentheses. Just explain optionalness later in the docstring.

adeak reviewed (commented):
My take on the style errors, cc @akaszynski

adeak commented on pyvista/plotting/charts.py:
```diff
@@ -0,0 +1,3375 @@
+
+        Examples
+        --------
+        >>> m, l, c = pyvista.Chart2D.parse_format("x--b")
```
And there's a doctest error coming from a missing import here:
```suggestion
        >>> import pyvista
        >>> m, l, c = pyvista.Chart2D.parse_format("x--b")
```
It would also be nice to say something about this example.  What does it exemplify?

dcbr commented:
So, I have been playing around with the sphinx documentation a bit. But I'm seeing some weird behaviour here and there. @akaszynski or @adeak could you please have a look and help me figure this out?

1. I modified `doc/api/plotting/index.rst` to list `Pen` and `Brush` objects correctly. This works, but throughout the whole documentation it seems I have to refer to these objects as `pyvista.plotting.charts.Pen`. If I use something else (e.g. `pyvista.charts.Pen`) it is not correctly recognized (no links created in the docs). That's however quite confusing, as trying to instantiate a new Pen using
    ```py
    import pyvista
    pen = pyvista.plotting.charts.Pen()
    ```
    results in an AttributeError. The correct way seems to be:
    ```py
    import pyvista
    pen = pyvista.charts.Pen()
    ```
    Could this be made consistent in the docs somehow?
2. I thought it would be nice to have an overview table for the different line styles, marker styles and color schemes you can use in plots. I started with the line styles in the `Pen` class. I created a script `make_chart_style_tables.py` which creates the necessary plots/images and produces an rst file with the table formatting. This rst file is then included in the `Pen` class' docstring. That seems to work fine (maybe the css of the table should be changed slightly to correctly align the cell contents), but I'm getting some sphinx warnings when building (for every image).
![image](https://user-images.githubusercontent.com/15089458/135762947-0ed9d507-63b5-4a0c-8b31-ac388a752120.png)
    ```
    WARNING: image file not readable: ../images/charts/ls_0.png
    ```
    Not sure what is causing that, as the images are displayed correctly in the resulting docs.
    In general, is this a good way to proceed (for the remaining marker style and color scheme tables) or should I create such tables differently?
3. There's also another numpydoc warning during building, probably related to the usage of the `_vtkWrapperMeta` class I'm using. Not sure how that can be resolved either.
    ```
    WARNING: [numpydoc] Validation warnings while processing docstring for 'pyvista.plotting.charts.Pen':
      PR01: Parameters {'_wrap', '**kwargs', '*args'} not documented
      PR02: Unknown parameters {'style', 'color', 'width'}
    ```
4. Many property examples are not properly shown in the docs, while others are. Any idea why that is happening?
![image](https://user-images.githubusercontent.com/15089458/135764163-43d1573f-3817-437a-b53f-7cfd2ea33fd0.png)

akaszynski commented:
I'd really like this merged fro 0.33.0, and it looks like we're still held up by:
```
File "<string>", line 2, in <module>
File "/home/vsts/work/1/s/pyvista/plotting/plotting.py", line 4881, in show
raise RuntimeError("This plotter has been closed and cannot be shown.")
RuntimeError: This plotter has been closed and cannot be shown.
WARNING: autodoc: failed to import class 'Pen' from module 'charts'; the following exception was raised:
No module named 'charts'
WARNING: autodoc: failed to import property 'Pen.color' from module 'charts'; the following exception was raised:
No module named 'charts'
WARNING: autodoc: failed to import property 'Pen.style' from module 'charts'; the following exception was raised:
No module named 'charts'
WARNING: autodoc: failed to import property 'Pen.width' from module 'charts'; the following exception was raised:
No module named 'charts'
WARNING: [numpydoc] Validation warnings while processing docstring for 'pyvista.plotting.charts.Brush':
PR01: Parameters {'_wrap'} not documented

WARNING: [numpydoc] Validation warnings while processing docstring for 'pyvista.ChartPie.show':
RT02: The first line of the Returns section should contain only the type, unless multiple values are being returned

WARNING: [numpydoc] Validation warnings while processing docstring for 'pyvista.ChartBox.show':
RT02: The first line of the Returns section should contain only the type, unless multiple values are being returned


Exception: Sphinx reported unique 12 warnings
```

dcbr commented:
Ok, I'm ready with this PR @akaszynski and @adeak. I will fix the last failing checks and then mark it as ready for review.
There's two remaining issues regarding the docs that I wasn't able to figure out:

- The Axis.log_scale example seems to use a wrong range for the x_axis.
- Boxplots are not properly drawn.

Not sure what is causing this, as these examples work just fine when manually executing them in an interactive console. Also not sure how I would be able to debug this?

akaszynski commented:
> Not sure what is causing this, as these examples work just fine when manually executing them in an interactive console. Also not sure how I would be able to debug this?

I'd not worry about it then.  This is already a gigantic PR and we can fix it in a follow up.  Would prefer closing this out before the next release.

akaszynski commented:
Also, ping me if you need help debugging this.  I'm planning a release in the next week or two since we've had several features added lately that I'd like incorporated.

dcbr commented:
The style checks are easy to fix (will do in a moment). The static checks as well, although I'm surprised this is necessary (specifying the type for class attributes), as I haven't done it for other classes either and there seems to be no complaints about that?
The documentation warnings are false and will be automatically fixed once numpy/numpydoc#337 is merged.

dcbr commented:
Hmm it seems I made the static checks worse now, I also cannot reproduce this on my local pc (running `mypy pyvista/core/ --no-incremental` yields "no issues found"). Could you have a look at this? I'll revert my previous changes.

dcbr commented:
I was able to reproduce the mypy errors using WSL on windows and 'fixed' them by adding ignore flags. That should have resolved all failing tests, but it seems there are a bunch of failing tests now that are unrelated to this PR?
The windows tests fail on the 'test_link_views' test, which have a decorator to skip them for VTK dev wheels. Could this be related to the new VTK wheels that were released last night?

akaszynski commented:
Failures are indeed due to the recent release of `vtk==9.1.0`.  Hotfix coming soon.

dcbr commented:
Ok great! I also think the docstringcheck failures are due to this new vtk release, as the `__repr__` method for vtk objects [was recently changed](https://gitlab.kitware.com/vtk/vtk/-/commit/fcc835940ec1c30594aded3f5c9dbe5651920f02).
So `pyvista.core.dataset.DataSet.textures` and `pyvista.core.dataset.DataSet.clear_textures` will have to change their expected outputs. Something like
```py
{'2k_earth_daymap': <Texture(...) at ...>}
```
maybe (untested and I'm not sure multiple ellipsis are allowed)?

akaszynski commented:
Both CI errors related to 9.1.0 have been fixed in #1788.  I'll have that merged tomorrow (after review), and then you're free to merge with `main` and then I'll finally be able to do a through review of this PR.

akaszynski commented:
Started making a few changes:

- Moved the charts API documentation into its own toctree.  It was making the API tree too cluttered.
- Renamed the fixtures to match PEP8 naming convention.
- Added label font size API.  I'm considering adding these to the theme class because the default chart font in our document looks quite small.

Onandon11 commented:
Any news on when this PR will be merged and thereafter released? It would be great to have this functionality 😁

dcbr commented:
The implementation is ready. I have been working on some VTK changes in the last two weeks, but those are not necessary for this PR to be merged (will update it later, once the VTK changes are released).
So once Alex and/or Andras find the time to review this, it should go rather quick :) But I think both of them are quite busy at the moment, so it might take a bit longer.

adeak commented:
Yeah, so the problem is two-fold: I don't know anything about VTK charts (never heard of the concept before this PR), and the PR itself is _crazy huge_. If I remove the image cache and whitespace-only diffs, there's still 20 changed files left with about 5k lines of diffs :cold_sweat:

I also wonder if this whole new section (with so many images generated) slows down doc building noticeably. I kicked CI, let's hope it won't segfault this time, then we can look at the build time.

dcbr commented:
I can totally understand that, it became much larger than I anticipated when I started working on this... So take your time to go through it and feel free to ask for clarifications.
As for the new generated images, I assume you are talking about the functions in `make_chart_style_tables.py`? I just did a quick test on my local pc and there it finishes in ~4.5s (to run the `make_all` function), so I don't think that would be the bottleneck.

Onandon11 commented:
Didn't mean to push or be inpatient, just eager to play with these new features, and replace them with my own (because these are way more mature).

If I can help with anything please just ask 😄 

PS. I noticed the new VTK release (9.1.0) broke some PyVista functionalities, don't know if they influence this PR tough...

adeak commented on doc/conf.py:
```diff
@@ -1,6 +1,10 @@
 import os
 import sys
 import datetime
+import locale
+
+# Otherwise VTK reader issues on some systems, causing sphinx to crash. See also #226.
+locale.setlocale(locale.LC_ALL, "en_US.UTF-8")
```
Just so I understand this correctly: VTK IO takes into account locales, so reading a file from a different locale can break. Since our vtk files for the docs are guaranteed to be in the `en_US` locale, this needs to be done not to break doc building on any locale.

dcbr replied:
Exactly, e.g. my own locale mixed up commas and dots, leading to the sphinx crash. Actually this is not really related to this PR (as it just broke doc building before the charts changes), but I forgot to open a separate issue for that. Should I do that instead and remove it from this PR?

adeak replied:
Nah, we (cough cough :laughing:) do that all the time with minor changes. Since this is confined to doc building and it just fixes an edge case bug, I think it's fine here.

akaszynski replied:
Good catch, because I completely missed this while debugging the sphinx build.

adeak commented on doc/conf.py:
```diff
@@ -9,8 +13,10 @@
 
 sys.path.insert(0, os.path.abspath("."))
 import make_external_gallery
+import make_chart_style_tables
 
 make_external_gallery.make_example_gallery()
+make_chart_style_tables.make_all()
```
If I understand correctly, there are three .rst pages that are generated on the fly, and these pages use the 70+ images added to version control with this PR.

If this is the case: do we really want to add those 70+ images to version control? What is the purpose of having them in git? Are we verifying them against the repo during generation, like with the plotting image cache, to catch regressions?

dcbr replied:
Yes.
And not sure if there are any benefits to adding them to git, I'm not doing anything special with them at the moment. I just added them as I saw some other images were already added to the repo as well.

adeak replied:
Yeah, those are all part of the [image cache](https://github.com/pyvista/pyvista/blob/e77405d0b97f2beeaf167ef5bb4dbfed619668e3/tests/plotting/test_plotting.py#L82-L146). Running tests for pyvista.plotting compares the output figures (just the VTK screenshots) to a "reference" copy that is in the image cache, tracked by git. When a figure changes a bit we get a warning during testing, and when a figure changes too much an error is raised. This signals that something drastically change the output. Either the change is reasonable, in which case the corresponding cache image must be invalidated and the new image `git add`ed, or something inadvertently broke that needs fixing.

If we don't expect to break charts in subtle ways then I we probably shouldn't add them to git. On the other hand, we could do the same image caching for the _test suite_ of charts.

dcbr replied:
All right, well the main intent to add these was for user convenience in the docs. I guess the testing is less important, as I don't really customize anything on the python side, so it would only break/change when things are modified on the VTK side (but in that case it might be handy to see a failure somewhere maybe?).

adeak replied:
Can you explain what you mean by "user convenience"? When the doc build is run, it will regenerate those images anyway, right? So them being in git would only have the effect of small unintentional changes showing up as diffs, which would then lead to the binary files being updated (and binary files are not great for git, but these are smaller files so that's less of a concern).

My point is that having these in git is probably less convenient for devs :) We could remove them from git and add them to .gitignore and devs wouldn't even notice they're being generated (unless someone has to touch the code that generates them).

dcbr replied:
User convenience for the users reading the docs, i.e. they immediately see in the docs what certain line/marker styles or color schemes look like.
I'm totally open for changing the way they are generated or stored though.

adeak replied:
Oh yeah, definitely keep the table :) I just meant moving stuff under the hood.

adeak commented on pyvista/core/objects.py:
```diff
@@ -47,10 +47,18 @@ def __init__(self, *args, **kwargs):
 
+    @staticmethod
+    def _prepare_arrays(arrays):
+        arrays = np.array(arrays, copy=False)
```
If this is meant to say "turn it into an array unless it's already an array", I'd prefer
```suggestion
        arrays = np.asarray(arrays)
```

adeak commented on pyvista/plotting/tools.py:
```diff
@@ -461,8 +461,53 @@ def opacity_transfer_function(mapping, n_colors, interpolate=True,
 def parse_color(color, opacity=None, default_color=None):
     """Parse color into a vtk friendly rgb list.
 
+    If color is a sequence of RGBA floats, opacity will be extracted
+    from ``color`` and not the ``opacity`` parameter.
```
If I understand this correctly (which I'm unsure of), I'd phrase this as
```suggestion
    If ``color`` is a sequence of RGBA floats, the ``opacity`` parameter
    is ignored.
```
(give or take line wrapping).

"sequence of RGBA floats" is also a bit misleading, but reading the parameter description below that should not be confusing.

dcbr replied:
Your understanding is correct, the opacity parameter is just a default to use when it is not present in the input (and hence forces the output to be an RGBA 4-tuple instead of an RGB 3-tuple).

adeak commented on pyvista/plotting/tools.py:
```diff
@@ -461,8 +461,53 @@ def opacity_transfer_function(mapping, n_colors, interpolate=True,
+    Returns
+    -------
+    tuple
+        Either a length 3 RGB sequence if opacity is unset, or RGBA
+        sequence when ``opacity`` is set.
```
Or when the input is also a sequence of RGBA values, right?

dcbr replied:
Exactly

adeak commented on pyvista/plotting/tools.py:
```diff
@@ -461,8 +461,53 @@ def opacity_transfer_function(mapping, n_colors, interpolate=True,
 def parse_color(color, opacity=None, default_color=None):
     """Parse color into a vtk friendly rgb list.
```
```suggestion
    """Parse color into a VTK friendly RGB(A) list.
```

adeak commented on pyvista/plotting/tools.py:
```diff
@@ -471,10 +516,10 @@ def parse_color(color, opacity=None, default_color=None):
-        pass
-    elif len(color) == 4:
-        color = color[:3]
+    elif isinstance(color, (list, tuple)) and len(color) in (3, 4) and np.all([isinstance(c, (int, float)) and 0 <= c <= 1 for c in color]):
```
1. This line is way too long.
2. I'd go for first an `isinstance(color, (collections.abc.Sequence, np.ndarray)` test, then inside call `np.asarray`, then checking the dtype and values for the last check. Combination of `np.issubdtype` and `((0 <= c) & (c <= 1)).all()`.

I think the current implementation breaks twice for numpy arrays: both `isinstance` checks.

adeak replied:
This is marked as resolved but I don't think this was fixed yet (or at least the fix hasn't been pushed).

dcbr replied:
Yea sorry, I was resolving them as I was fixing them locally to keep an overview of which things I still had to address. With the latest push everything should have been addressed :)

adeak commented on doc/api/plotting/charts/index.rst:
```diff
@@ -0,0 +1,25 @@
+   :toctree: _autosummary
+   :template: custom-class-template.rst
+
+   plotting.charts.Pen
+   plotting.charts.Brush
+   plotting.charts.Axis
+   plotting.charts.LinePlot2D
+   plotting.charts.ScatterPlot2D
+   plotting.charts.BarPlot
+   plotting.charts.AreaPlot
+   plotting.charts.StackPlot
+   Chart2D
+   plotting.charts.BoxPlot
+   ChartBox
+   plotting.charts.PiePlot
+   ChartPie
+   ChartMPL
```
I haven't had time to run this PR's build yet. How do these guys get rendered? I wonder if we should separate them according to "pulled into main `pyvista` namespace" vs "`pyvista.charts` only".

dcbr replied:
Yes that's possible, the current order is just the order in `charts.py`. I will add a screenshot of the docs here tomorrow.

dcbr replied:
Yes that's possible, the current order is just the order of definition in `charts.py`. I will add a screenshot of the docs here.

dcbr replied:
Seems something is broken in the sidebar of the docs, will investigate that tomorrow.
Charts overview page:
![image](https://user-images.githubusercontent.com/15089458/143132347-fca1aa11-ab0f-45b2-8f16-b62686c81bb5.png)
Pen class (sidebar is gone here):
![image](https://user-images.githubusercontent.com/15089458/143132426-7d858b2d-5767-481b-9a39-19bf4bb74adf.png)

adeak replied:
Awesome, thank you.

So it's clear that the namespace differences are obvious on the main Charts page. I'd pull the four exceptions to the front, and maybe even write a paragraph before the class list that most of the chart classes are in `pyvista.plotting.charts`, with the exceptions of [the classes] for convenience reasons. Or something along those lines.

We'd want people to use the short names where available, but at the same time they should know that there's treasure two namespaces deeper.

dcbr replied:
Ok, will do that. As an extra, do you know why sphinx uses `pyvista.plotting.charts.[class]` while the correct way to access these classes in python is `pyvista.charts.[class]` and is there a way to make this consistent?

adeak replied:
I'm afraid I'm a sphinx/numpydoc noob, so I don't know how to fix that automatically. You can probably define them manually with a short name.

A related example that comes to mind is [`pv.Light`](https://dev.pyvista.org/api/core/lights.html):
```py
>>> import pyvista as pv
>>> type(pv.Light())
<class 'pyvista.plotting.lights.Light'>
```

And its API reference resolves
```
:class:`pyvista.Light`
```
just fine. But that case might be a bit different, because we don't have such a list of explicit classes for Lights...

dcbr replied:
I just pushed part 2 of the changes, this should have addressed your remaining suggestions. The docs are also correctly building now.

adeak reviewed (commented):
Just a first quick glance to get the ball moving...

adeak commented on pyvista/plotting/tools.py:
```diff
@@ -485,7 +530,7 @@ def parse_color(color, opacity=None, default_color=None):
         color='#FFFFFF'""")
     if opacity is not None and isinstance(opacity, (float, int)):
         color = [color[0], color[1], color[2], opacity]
```
If we claim that this returns a tuple, we should return a tuple. And numpy dtypes are still excluded here, which is a problem for numpy array valued input.

dcbr replied:
I'm working on that right now, will push part 2 once I'm ready :)
A tuple is already returned right?

adeak replied:
Ah, sorry, I misread, I thought we were returning this `color` list. Yes, ignore the tuple part then; sorry!

adeak commented on pyvista/plotting/tools.py:
```diff
@@ -459,23 +460,78 @@ def opacity_transfer_function(mapping, n_colors, interpolate=True,
 
 def parse_color(color, opacity=None, default_color=None):
-    """Parse color into a vtk friendly rgb list.
+    """Parse color into a VTK friendly RGB(A) list.
```
```suggestion
    """Parse color into a VTK friendly RGB(A) tuple.
```

adeak commented on pyvista/plotting/tools.py:
```diff
@@ -459,23 +460,78 @@ def opacity_transfer_function(mapping, n_colors, interpolate=True,
 
+    Parameters
+    ----------
+    color : str or sequence
+        Either a string, RGB sequence, RGBA sequence, or hex color string.
+        For example:
+
+            * ``'white'``
+            * ``'w'``
+            * ``[1, 1, 1]``
+            * ``[0.5, 1, 0.7, 1]``
+            * ``'#FFFFFF'``
```
The implementation seems to assume float values between 0 and 1. We should probably explicitly document this, because users might expect uint8 values to be handled correctly.

dcbr replied:
Yea, I was actually thinking of creating a proper `Color` class with all necessary `from_*` and `to_*` conversion functions, but that would be too much out of scope for this PR. So for now I will just document it indeed.

adeak commented on pyvista/plotting/tools.py:
```diff
@@ -459,23 +460,78 @@ def opacity_transfer_function(mapping, n_colors, interpolate=True,
-        color = color[:3]
+    elif isinstance(color, (Sequence, np.ndarray)):
+        try:
+            color = np.asarray(color, dtype=np.float64)
```
Why the explicit conversion to double?

dcbr replied:
That replaces the dtype check. If the conversion fails (non numeric data provided as input), we catch the ValueError and raise the more user friendly ValueError below.

dcbr replied:
Ah, but that might create an unwanted copy I guess...

dcbr replied:
Although it's probably best to have consistent float/double return values maybe?

adeak replied:
Yeah, the return type is best if type consistent. This can indeed cause potential copies, but we're talking about really small arrays. I wouldn't worry about it.

adeak commented on pyvista/plotting/tools.py:
```diff
@@ -459,23 +460,78 @@ def opacity_transfer_function(mapping, n_colors, interpolate=True,
+    elif isinstance(color, (Sequence, np.ndarray)):
+        try:
+            color = np.asarray(color, dtype=np.float64)
+            if len(color.shape) != 1 or color.shape[0] not in (3, 4) or not np.all((0 <= color) & (color <= 1)):
```
```suggestion
            if color.ndim != 1 or color.size not in (3, 4) or not np.all((0 <= color) & (color <= 1)):
```

adeak commented on pyvista/plotting/tools.py:
```diff
@@ -459,33 +460,95 @@ def opacity_transfer_function(mapping, n_colors, interpolate=True,
+        if isinstance(opacity, (float, int)) and 0 <= opacity <= 1:
+            color = [color[0], color[1], color[2], float(opacity)]
+        else:
+            raise ValueError(f"""
+    Invalid opacity input: {opacity}
+    Must be a scalar value between 0 and 1.""")
```
Now this is really just nitpicking, but we usually handle multiline error messages with implicit line continuations:
https://github.com/pyvista/pyvista/blob/0c344920d6959205aaa251bdb190b05c1487a1c1/pyvista/core/filters/data_set.py#L4341-L4344
or sometimes with `textwrap.dedent`:
https://github.com/pyvista/pyvista/blob/0c344920d6959205aaa251bdb190b05c1487a1c1/pyvista/plotting/plotting.py#L4778-L4783

I personally dislike letting strings "underflow" into lower indentation levels.

akaszynski commented:
#1871 will make merging this a bit difficult. @banesullivan, would you mind taking a look at this. If you don't get to it, I'll take a look at it tomorrow.

banesullivan commented:
I can take a look tomorrow afternoon/evening and help!

dcbr commented:
I'm having some weird doc building issues. When I do a clean build, all chart screenshots (from example code in the docstrings) have a size of [400, 400] instead of the usual [1024, 768] (from the global theme). This screws up some chart examples (too small, overlapping labels). However when I only rebuild the chart docs everything is fine (screenshots have normal size). Could there be some 'leak' from a previous docstring example that overwrites the default theme settings for the remaining docstring examples?

akaszynski commented:
> I'm having some weird doc building issues. When I do a clean build, all chart screenshots (from example code in the docstrings) have a size of [400, 400] instead of the usual [1024, 768] (from the global theme). This screws up some chart examples (too small, overlapping labels). However when I only rebuild the chart docs everything is fine (screenshots have normal size). Could there be some 'leak' from a previous docstring example that overwrites the default theme settings for the remaining docstring examples?

I'm seeing some examples that actually change the global theme.  Let me push a fix and let me know if this fixes it.

banesullivan commented on tests/plotting/test_plotting.py:
```diff
@@ -15,6 +18,7 @@
 import numpy as np
 import pytest
 import vtk
+import matplotlib.pyplot as plt
```
Matplotlib is currently a soft requirement

banesullivan commented:
@dcbr, I just merged main into this and did a bit of reformatting. This is a very large diff so I apologize if I messed anything up. I also did what I could to make sure some of the new CI checks are happy

banesullivan commented on AUTHORS.rst:
```diff
@@ -25,3 +25,4 @@ considered "The PyVista Developers":
 - Eric Larson, (`@larsoner <https://github.com/larsoner>`_)
 - Matthew Flamm, (`@MatthewFlamm <https://github.com/MatthewFlamm>`_)
 - Darik Gamble, (`@darikg <https://github.com/darikg>`_)
+- dcbr, (`@dcbr <https://github.com/dcbr>`_)
```
@dcbr, if you are comfortable sharing your name, please add update it here. I think this massive of a contribution immediately warrants your name on the collective work! cc @pyvista/developers

dcbr replied:
Thanks a lot :D

dcbr commented:
Latest commit should fix the mypy errors (was due to my own changes in the latest commit). Thanks for merging with main and fixing the conflicts, everything seems to work fine :)
@akaszynski I will rebuild the docs and check if those issues are resolved now!

dcbr commented:
I can't wrap my head around this one...
When rebuilding the docs for the charts module only (by deleting the `doc/api/plotting/charts/_autosummary` folder), I get below correct plot used in the stackplot docs.
![stackplot_right](https://user-images.githubusercontent.com/15089458/144705296-f2291127-f03e-4144-bc57-186be4434092.png)
When rebuilding the docs for the charts module, `pyvista.Plotter.disable_parallel_projection` AND `pyvista.Plotter.enable_hidden_line_removal` (by deleting their respective rst files from the `doc/api/plotting/_autosummary`
 folder), I get below wrong plot used in the stackplot docs (the y axis label is not properly oriented and not in bold, the plot's width seems off to).
![stackplot_wrong](https://user-images.githubusercontent.com/15089458/144705378-c30c1c06-0a69-4da3-8c9a-497bd3631562.png)
I also tried to reproduce this issue using the following script (as the docs take quite a long time to build), but that didn't work either (the correct plot is produced). So I'm out of ideas as to why this is happening...
```py
import pyvista
from pyvista.ext import plot_directive

pyvista.OFF_SCREEN = True
pyvista.BUILDING_GALLERY = True
pyvista.set_plot_theme("document")
pyvista.global_theme.window_size = [1024, 768]
pyvista.global_theme.font.size = 22
pyvista.global_theme.font.label_size = 22
pyvista.global_theme.font.title_size = 22
pyvista.global_theme.return_cpos = False
pyvista.set_jupyter_backend(None)


methods = [
    pyvista.Plotter.disable_parallel_projection,
    pyvista.Plotter.enable_hidden_line_removal,
    pyvista.charts.BarPlot
]
context = True

for method in methods:
    plot_directive.render_figures(method.__doc__, None, "P:/projects/tmp_imgs", method.__name__, context, None, None)
```
At this point I just hope the fix in #1912 will magically fix this :)

dcbr commented:
Unfortunately #1915 did not fix the chart label issue :/
I did find a minimal failing example though. Plotting a chart, followed by a plot with custom `add_text`, followed by a chart with a different y label, causes the last chart's label to be messed up (see code and plots below). Unfortunately I don't think I'll be able to easily fix it myself as it's probably a VTK/OpenGL issue (because the vtkTextProperty of the axis title is correctly set at all times).
<details><summary>Code to reproduce</summary>

```py
import pyvista

SCREENSHOT = False
if SCREENSHOT:
    pyvista.OFF_SCREEN = True
    pyvista.BUILDING_GALLERY = True
pyvista.set_plot_theme("document")


def plot_chart(label):
    pl = pyvista.Plotter()
    chart = pyvista.Chart2D()
    chart.line([0, 1, 2], [2, 1, 3])
    chart.y_axis.label = label
    pl.add_chart(chart)
    pl.show()


def screenshot(name):
    if SCREENSHOT:
        pl = next(iter(pyvista.plotting._ALL_PLOTTERS.values()))
        pl.screenshot(name)
        pyvista.close_all()


plot_chart("Label before")
screenshot("tmp_imgs/chart_before.png")

pl = pyvista.Plotter()
pl.add_text("Foo")
pl.show()
screenshot("tmp_imgs/plotter_text.png")

plot_chart("Custom after")
screenshot("tmp_imgs/chart_after.png")
```
</details>

<details><summary>Before and after plots</summary>
<table>
<tr>
<th>
Before chart
</th>
<th>
After chart
</th>
</tr>
<tr>
<td>

![chart_before](https://user-images.githubusercontent.com/15089458/144917576-6c4741b1-323d-42d3-9792-6e46059663d5.png)
</td>
<td>

![chart_after](https://user-images.githubusercontent.com/15089458/144917617-a8b4c5a1-0612-4cb9-a85d-c33882dbac7f.png)
</td>
</tr>
</table>
</details>

adeak commented:
I can reproduce the issue using your example. On linux, both vtk 9.1.0 and 9.0.3.

Onandon11 commented:
Just out of curiosity when will then next release be and will this be included?

akaszynski commented on doc/make_chart_style_tables.py:
```diff
@@ -0,0 +1,142 @@
+    p.background_color = 'w'
+    chart = pv.Chart2D()
+    tmp_plot = chart.bar([0], [[1]]*2, color=color_scheme, orientation="H")
+    N = len(tmp_plot.colors)  # Use a temporary plot to determine the total number of colors in this scheme
+    plot = chart.bar([0], [[1]]*N, color=color_scheme, orientation="H")
+    chart.remove_plot(tmp_plot)
+    plot.pen.color = 'w'
+    chart.x_range = [0, N]
+    chart.hide_axes()
+    p.add_chart(chart)
+    _, img = p.show(screenshot=True, return_cpos=True)
+    # Crop the image and save it
+    p._save_image(img[34:78, 22:225, :], path, False)
+    return N
```
Nitpick, case of parameter.
```suggestion
    n_colors = len(tmp_plot.colors)  # Use a temporary plot to determine the total number of colors in this scheme
    plot = chart.bar([0], [[1]]*n_colors, color=color_scheme, orientation="H")
    chart.remove_plot(tmp_plot)
    plot.pen.color = 'w'
    chart.x_range = [0, n_colors]
    chart.hide_axes()
    p.add_chart(chart)
    _, img = p.show(screenshot=True, return_cpos=True)
    # Crop the image and save it
    p._save_image(img[34:78, 22:225, :], path, False)
    return N
```

akaszynski commented on doc/make_chart_style_tables.py:
```diff
@@ -0,0 +1,142 @@
+     - .. image:: /{}
+"""
+    img_path = f"{IMAGE_DIR}/cs_{i}.png"
+    N = make_color_scheme_img(token, img_path)
+    return row_template.format(token, data["descr"], N, img_path)
```
```suggestion
    n_colors = make_color_scheme_img(token, img_path)
    return row_template.format(token, data["descr"], n_colors, img_path)
```

akaszynski commented:
> Just out of curiosity when will then next release be and will this be included?

Next release by EOY, and this will be included.

@pyvista/developers, can we call this good to go? I'm doing one final review now, but as it is I think it's already sufficient for release.

akaszynski commented:
> Note: Only tested on my local pc (Windows 10, python 3.7, vtk 9.0.3)

Tested on Linux (Ubuntu 20.04) with Python 3.8 9.1.0 with no issues.

Going to approve and merge unless there are any major objections. I think this will be a great addition to the library!

adeak commented:
I probably won't have time to give this a thorough read in the next week so don't hold back on my account.

dcbr commented:
Great :)
I also found what caused the doc building issue and am working on an upstream (vtk) fix. So once vtk 9.2 releases, all currently known issues with the docs should be resolved (and some internal code pieces can be greatly simplified) :)

akaszynski commented:
> I probably won't have time to give this a thorough read in the next week so don't hold back on my account.

Recommending a followup PR in that case.  It's a long enough PR as it is and I want to make sure that it gets in for the release.

> once vtk 9.2 releases, all currently known issues with the docs should be resolved (and some internal code pieces can be greatly simplified)

We're probably going to want to keep that in for backwards compatibility. We still have parts in for `vtk==8.1.2` that will probably stay in for quite some time.

akaszynski reviewed (approved):
This has been sufficiently reviewed to be a feature. We can have a follow-up PR for any further changes, but I feel the overall design is sound and it doesn't affect the rest of the library.
